### ISSUE TITLE
clarify the generation and use of All.v files

### ISSUE THREAD

**Issue by rmatthes** [UniMath/UniMath#1589]
Triggered by PR #1588 (but could have started through PR #1583), we should have a clear policy about which All.v files to create and when to allow their use not just for the purposes of checking consistency or rapid prototyping but also in developments (incl. the satellite repos).
Currently, there is one All.v per package of UniMath and an All.v at the root. They are auto-generated. There are also hand-crafted All.v in two subdirectories of packages: in UniMath/CategoryTheory/Chains and CategoryTheory/categories/HSET. I already mention that they do not list all .v files of their respective directories, contrary to what their name suggests (this is due to recent merges). Hand-crafted files easily suffer this problem, so this would ask for another consistency check in CI (which should be easy in comparison with other already implemented checks).

Importing an auto-generated All.v breaks the current proposal of a dune installation (PR #1588). Therefore, that PR suggests to replace the auto-generated UniMath/MoreFoundations/All.v and UniMath/Foundations/All.v by hand-crafted ones. This does not solve the problem of import of UniMath/CategoryTheory/All.v in some file in TypeTheory.

And this raises the question if we should not have more All.v in subdirectories of packages and disallow imports of All.v of too big chunks of UniMath.

**Comment by nmvdw** [comment]
> And this raises the question if we should not have more All.v in subdirectories of packages and disallow imports of All.v of too big chunks of UniMath.

This is actually a very valid question, and I think that we should handcraft more `All.v` files. Why? Well, do we really want this many `Require Import`s in any file: https://github.com/UniMath/UniMath/blob/master/UniMath/Bicategories/OtherStructure/Examples/StructureBicatOfUnivCats.v#L12? A follow-up question would be whether we can find some guidelines about which `All.v`-files to generate.

I also agree with removing (some) auto-generated `All.v`-files, such as `UniMath/CategoryTheory/All.v` and `UniMath/Bicategories/All.v`. These create unnecessary dependencies, and I think that's not a good idea for the packages that take a longer time to compile.

**Comment by m-lindgren** [comment]
There's also the option of still generating the `All.v`-files but having them under version control, like what is done currently for `CONTENTS.md`. I don't know what I prefer. I wouldn't mind if files named `All.v` actually exports _all_ of the stuff in that folder, because I would be confused if they didn't. And in that case it's easier to have them generated.

I agree that It would be good to have a clear policy with regard to adding unnecessary includes. But I'm not sure about removing the large ones, I mean would they not be useful during development for `Search` and so on? I don't know, I don't use them, but that's mostly because I haven't thought of it.

**Comment by rmatthes** [comment]
I agree that another use of `All.v` besides checking consistency and rapid prototyping is searching. So, we should rather ban the use of `All.v` of huge packages than deleting those files from the repo.

**Comment by rmatthes** [comment]
Concerning `CONTENTS.md`, this is sometimes irritating enough that it has to be committed so often. Having dozens of such files does not look good for me.

**Comment by DanGrayson** [comment]
If there is a hand-crafted All.v file, then it shouldn't be called All.v, if it doesn't include all the files.  And if it includes all the files, then it might as well be automatically generated.
